chia/full_node/full_node_api.py

```diff
@@ -1720,5 +1712,201 @@ def get_fee_estimates(est: FeeEstimatorInterface, req_times: List[uint64]) -> Li
+    ) -> Message:
+        peer_id = peer.peer_node_id
+        max_subscriptions = self.max_subscriptions(peer)
+        subs = self.full_node.subscriptions
+
+        peak_height = self.full_node.blockchain.get_peak_height()
+        assert peak_height is not None
+
+        header_hash = self.full_node.blockchain.height_to_hash(peak_height)
+        assert header_hash is not None
+
+        new_subscriptions = set(request.coin_ids) - subs.coin_subscriptions(peer_id)
+        added = subs.add_coin_subscriptions(peer_id, list(new_subscriptions), max_subscriptions)
+
+        response = wallet_protocol.RespondAddCoinSubscriptions(list(added), peak_height, header_hash)
+        msg = make_msg(ProtocolMessageTypes.respond_add_coin_subscriptions, response)
+        return msg
+
+    @api_request(
+        peer_required=True,
+        reply_types=[ProtocolMessageTypes.respond_remove_puzzle_subscriptions],
+    )
+    async def request_remove_puzzle_subscriptions(
+        self, request: wallet_protocol.RequestRemovePuzzleSubscriptions, peer: WSChiaConnection
+    ) -> Message:
+        peer_id = peer.peer_node_id
+        subs = self.full_node.subscriptions
+
+        if request.puzzle_hashes is None:
+            removed = list(subs.puzzle_subscriptions(peer_id))
+            subs.remove_puzzle_subscriptions(peer_id, removed)
+        else:
+            removed = list(subs.remove_puzzle_subscriptions(peer_id, request.puzzle_hashes))
+
+        response = wallet_protocol.RespondRemovePuzzleSubscriptions(removed)
+        msg = make_msg(ProtocolMessageTypes.respond_remove_puzzle_subscriptions, response)
+        return msg
+
+    @api_request(
+        peer_required=True,
+        reply_types=[ProtocolMessageTypes.respond_remove_coin_subscriptions],
+    )
+    async def request_remove_coin_subscriptions(
+        self, request: wallet_protocol.RequestRemoveCoinSubscriptions, peer: WSChiaConnection
+    ) -> Message:
+        peer_id = peer.peer_node_id
+        subs = self.full_node.subscriptions
+
+        if request.coin_ids is None:
+            removed = list(subs.coin_subscriptions(peer_id))
+            subs.remove_coin_subscriptions(peer_id, removed)
+        else:
+            removed = list(subs.remove_coin_subscriptions(peer_id, request.coin_ids))
+
+        response = wallet_protocol.RespondRemoveCoinSubscriptions(removed)
+        msg = make_msg(ProtocolMessageTypes.respond_remove_coin_subscriptions, response)
+        return msg
+
+    @api_request(peer_required=True, reply_types=[ProtocolMessageTypes.respond_puzzle_state])
+    async def request_puzzle_state(
+        self, request: wallet_protocol.RequestPuzzleState, peer: WSChiaConnection
+    ) -> Message:
+        max_items = self.max_subscribe_response_items(peer)
+        max_subscriptions = self.max_subscriptions(peer)
+
+        previous_header_hash = (
+            self.full_node.blockchain.height_to_hash(request.previous_height)
```

Comment:
I'm hoping that we can phase out the `height_to_hash` function/class. It's not sustainable to keep this in RAM in the long run, and the DB v2 supports this lookup.

However, the closest thing to this is `prev_block_hash(height)`, but it's not quite what you need. I think it would be good to make an `async` version of `height_to_hash()` that looks it up in the DB instead. But maybe putting that in here is scope-creep

Reply:
I assumed this was doing a db lookup as is. Maybe we make this change in a separate PR I can rebase it into this one and switch how it's implemented here. Or we can do it after?

Reply:
I agree on separate PR. a function performing a db lookup has to be `async`. That's the main hurdle for transitioning.